File: 3_lab/prob_1/plugins/parrot.py
class Parrot:
    def __init__(self, name):
        self.p_name = name

    def name(self):
        return self.p_name

    def greet(self):
        return "ciao"

    def menu(self):
        return "orahe"


def create(name):
    return Parrot(name)

File: 3_lab/prob_1/parrot.c
#include <stdio.h>
#include <stdlib.h>

typedef char const *(*PTRFUN)();

typedef struct Parrot {
    PTRFUN* vtable;
    char const* name;
} Parrot;


char const* greet(void) {
  return "ciao!";
}

char const* menu(void){
  return "orahe";
}

char const* name(void* this) {
  return ((Parrot *)(this))->name; 
}

PTRFUN vTable[3] = {
    (PTRFUN)name,
	(PTRFUN)greet,
	(PTRFUN)menu,
};

void* create(char const* name) {
    Parrot* parrot = (Parrot *)malloc(sizeof(Parrot));
    parrot->vtable = vTable;
    parrot->name = name;
}